Add more flexibility in our ability to specify codeservers

ChrisPenner opened the pull request:
# Overview

Currently there are two different ways to specify your codeserver;

* `CodeServers` in `.unisonConfig`, then specify an alias to `auth.login`
* Provide the `UNISON_SHARE_HOST` env var when running unison.

This deletes the `CodeServers` config option and removes the `auth.login` argument.

We also had the issue that it's not straight-forward to provide a port, http scheme, or additional path to the specified share server, this is important because:

* The share api is currently hosted at the `/api/` prefix on share (for now)
* We need to be able to run locally, which means overriding the scheme to http; and specifying a port number.
* We still need a way to canonically identify each host such that we can easily look up the credentials we need for that host from the stored credentials file.

## Implementation

* The rule for finding share server endpoint is now:
  * Take the value of `UNISON_SHARE_HOST`, or `https://share-next.us-west-2.unison-lang.org/api` as a default value, 
  * Append `/.well-known/openid-configuration` and fetch the discovery document
  * Use the routes provided in the discovery document to authenticate with that host.
  * After authenticating, store the credentials under the `uriRegName` from the provided URI in our credentials file; (E.g. `share.unison-lang.org` for `https://share.unison-lang.org/api`.
  * When making an authenticated request, check if we have credentials for the request's `uriRegName`, and if so, attach those credentials.

ChrisPenner committed:
Add explicit types for share host and uri

ChrisPenner committed:
Ensure we identify servers correctly

ChrisPenner committed:
Remove ability to pass args to auth.login

and remove logic for CodeServers config

aryairani commented:
Thanks @ChrisPenner.

What's the stored credentials file, is that something we control?

I think we want to key off of the full URI and not just the `uriRegName`, unless there's some reason that's impossible? i.e. If we're supporting arbitrary URLs, then there could two independent servers with the same `uriRegName`.

aryairani commented:
Maybe not the full URI, but the scheme + regName + port + path

mitchellwrosen commented on unison-cli/src/Unison/Share/Types.hs:
```diff
@@ -0,0 +1,57 @@
+import Network.URI
+import Unison.Prelude
+
+newtype CodeserverURI = CodeserverURI URI
```
How is `CodeserverURI` different than `URI`? I'm wondering whether it has some additional structure that `URI` doesn't, but I also see there's a parse function `codeserverIdFromURI`.

mitchellwrosen replied:
Oh, I did a little more digging around the code and see that it is the thing that comes from `UNISON_SHARE_HOST`.

What do you think about making it instead a type like this, basically a `URI` that has a non-Maybe `uriAuth`?

```haskell
data CodeserverURI = CodeserverURI {
  uriScheme :: String,
  uriAuthority :: URIAuth,
  uriPath :: String,
  -- Arya says not these next two
  uriQuery :: String,
  uriFragment :: String
}
```

Then, we wouldn't need to later parse with `codeserverIdFromURI`, but we would have an additional way that parsing the top-level `CodebaseURI` can fail. (It can already fail with `error` if the string provided isn't a valid `URI`).

aryairani replied:
We also don't want `uriUserInfo` here (part of `uriAuth`) since that doesn't fit into the `BaseUrl` that the Servant client functions that our servant api generates anyway.

mitchellwrosen commented on unison-cli/src/Unison/Share/Types.hs:
```diff
@@ -0,0 +1,57 @@
+newtype CodeserverURI = CodeserverURI URI
+  deriving newtype (Show, Eq, Ord)
+
+instance Configurator.Configured CodeserverURI where
```
Do we still need this instance?

mitchellwrosen commented on unison-cli/src/Unison/Share/Types.hs:
```diff
@@ -0,0 +1,57 @@
+newtype CodeserverId = CodeserverId {codeserverId :: Text}
+  deriving newtype (Show, Eq, Ord)
+
+instance ToJSON CodeserverId where
```
Suggestion: newtype-derive these?

mitchellwrosen reviewed (approved):
LGTM, just one question about the `CodebaseURI` newtype, and whether it would be better with more structure.

aryairani reviewed (approved):
LGTM, though I think we should add that additional info to `CodeserverId`, since `uriRegName` won't be enough in general, but it's fine for the immediate term.

ChrisPenner commented:
@aryairani
> since uriRegName won't be enough in general, but it's fine for the immediate term.

I understand that anything could happen, but this seems pretty unlikely to me that distinct codeservers would be hosting on the same host AND that they have completely different authorization setups 🤔  (What I have implemented can handle multiple servers on the same host if they share Authentication tokens). I added port to the CodeserverId so we can have distinct servers on localhost; but can you explain when you think we might have codeservers with different authentication at different paths?

I also don't think we should add scheme, IMO scheme has to do with the transport layer, not Identity, and it seems like just another accidental mismatch waiting to happen and/or frustrate someone, especially when it comes to working on localhost.

It's possible but a little more work to support entire paths, because the authenticated HTTP client has to match each call to its credentials, so every URL needs a trivial way to convert into the 'key' we use to look up credentials for that host. 

So, if we support arbitrary URLs, how do we know that the key for `https://share-next.unison-lang.org/api/users/chrispenner` is `share-next.unison-lang.org/api`? There's no way to know in general, we could iterate through all of our keys and check path prefixes or something, but that seems clunkier. We can add it in the future if needed, but I suspect we won't ever need it.

ChrisPenner committed:
PR feedback: special structure for coderserver URI

ChrisPenner committed:
Add port to CodeserverId
